### ISSUE TITLE
Design `Session` type with extension plan

### ISSUE THREAD

**Issue by marsella** [boltlabs-inc/key-mgmt#30]
The current client API uses the `Session` type to generalize over network communications. In a normal usage flow, the caller will create a new `Session`, then pass it to API calls until they are finished or the session closes. The `Session` abstracts over network communication with the key servers.

This abstraction hasn't been thought through very thoroughly, though, and it's not clear to me that it can handle the complexity of the use cases we're trying to put it up to. For example:
- Some actions require communication with most or all the key servers (key creation) but others only need a threshold (signing)
- Server failures are independent, but we need to avoid cascading failures, like if one server times out, then the `Session` re-establishes that connection, but meanwhile other servers have timed out.
- Given a misbehaving key server (e.g. repeatedly closing the connection), some operations should still work. 

The `Session` will need some `pub(crate)` functionalities, too, to actually access the connections and send/receive over them. How do we want to give access to this? In some cases, it's irrelevant which key server gets a set of bytes (e.g. sending a secret share to a server). In other cases, there might be multi-message protocols to execute in parallel with each server, and the messages for one server can't be mixed up with those for another.

There may be two types of authentication. For the local client, we expect to use a PAKE-based scheme, which will require periodic "refreshing" if too much time elapses and the user needs to re-enter their password. For the remote client, we expect some kind of entity-based authentication that might be fully described in config files, not require any manual human input, and be reusable for requests on multiple different keys & users.

So, there are two components of the design:
1. The external interface: how do calling applications set up, reuse, and handle errors over secure communication channels for multi-operation flows? 
2. The internal interface: how does our software execute multi-message protocols with multiple parties over secure communication channels?

For Phase 1, we only need to handle communication with a single key server. However, it seems worth it to consider some options for extending the interface for multiple servers.

This issue is done when
- [ ] There's a clear plan for (1) and (2)
- [ ] It's implemented for single key servers
- [ ] The plan for extending it to multiple key servers is documented in the comments here.

**Comment by vidyaap** [comment]
This issue looks like it has aspects related to both application-layer and transport-layer session handling. It would be worth creating another issue about transport-layer aspects and separating out transport-layer aspects from here to the new issue.  The is related to https://github.com/boltlabs-inc/key-mgmt-spec/issues/4

**Comment by picklenerd** [comment]
Closing in favor of #175. I tried to re-state the concerns of this issue in terms of the new design built around `tonic`.
